src/modules/adaptors/Curve/CurveHelper.sol

```diff
@@ -0,0 +1,233 @@
+        if (useUnderlying) {
+            finalArgOrEmpty = ",bool";
+        }
+
+        return
+            bytes4(
+                keccak256(
+                    abi.encodePacked(
+                        "add_liquidity(uint256[",
+                        numberOfCoins.toString(),
+                        "],",
+                        "uint256",
+                        finalArgOrEmpty,
+                        ")"
+                    )
+                )
+            );
+    }
+
+    function _curveRemoveLiquidityEncodedCalldata(
+        uint256 lpTokenAmount,
+        uint256[] memory orderedTokenAmounts,
+        bool useUnderlyings
+    ) internal pure returns (bytes memory callData_) {
+        bytes memory finalEncodedArgOrEmpty;
+        if (useUnderlyings) {
+            finalEncodedArgOrEmpty = abi.encode(true);
+        }
+
+        return
+            abi.encodePacked(
+                _curveRemoveLiquidityEncodeSelector(orderedTokenAmounts.length, useUnderlyings),
+                lpTokenAmount,
+                abi.encodePacked(orderedTokenAmounts),
+                finalEncodedArgOrEmpty
+            );
+    }
+
+    /// @dev Helper to encode selector for a call to remove liquidity on Curve
+    function _curveRemoveLiquidityEncodeSelector(
+        uint256 numberOfCoins,
+        bool useUnderlyings
+    ) internal pure returns (bytes4 selector_) {
+        string memory finalArgOrEmpty;
+        if (useUnderlyings) {
+            finalArgOrEmpty = ",bool";
+        }
+
+        return
+            bytes4(
+                keccak256(
+                    abi.encodePacked(
+                        "remove_liquidity(uint256,",
+                        "uint256[",
+                        numberOfCoins.toString(),
+                        "]",
+                        finalArgOrEmpty,
+                        ")"
+                    )
+                )
+            );
+    }
+
+    // TODO can we add a check in this adaptor, that verifies the calling Cellar has `sharePriceOracle` in it?
+    // So if we put it in `balanceOf`, that would full stop all use of a non oracle cellar that adds it
+    // But recovering from this would require us to call forcePositionOut, otherwise the cellar is bricked
+    // I think in balance of if I can return zero before doing the check, then it would be fine.
+    // An attacker would be able to send Curve LP to the cellar to temporaily brick it, but
+    // if a lot of mistakes were made, and the position wasa added, then the second total assets call in callOnAdaptor would revert, and prevent
+    // a strategist from moving money into it
+    function _ensureCallerUsesOracle(address caller) internal view {
+        // Try calling `sharePriceOracle` on caller.
+        CellarWithOracle(caller).sharePriceOracle();
+    }
+
+    // TODO should this really restrict what can be called?
+    function _callReentrancyFunction(CurvePool pool, bytes4 selector) internal {
```

Comment:
Yes, though we really need to test every pools reentrancy function call because curve will randomly decide to not add a reentrancy check to a function that normally has it. It baffles me. But I also haven't found a way to write reentrancy tests for normal ERC20 token LP pairs.